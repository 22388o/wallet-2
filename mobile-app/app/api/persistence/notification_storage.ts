import AsyncStorage from '@react-native-async-storage/async-storage'

const KEY = 'WALLET.ENABLED_NOTIFICATION'

export enum NotificationType {
  TRANSACTION = 'TRANSACTION',
}

export interface AppNotificationTypesI {
  type: NotificationType
  value: boolean
  displayName: string
}

async function set (enabledNotifications: NotificationType[] = []): Promise<void> {
  await AsyncStorage.setItem(KEY, JSON.stringify(enabledNotifications))
}

async function get (): Promise<NotificationType[]> {
  const notificationTypes = await AsyncStorage.getItem(KEY) ?? '[]'
  return JSON.parse(notificationTypes)
}

export const NotificationPersistence = {
  set,
  get
}
