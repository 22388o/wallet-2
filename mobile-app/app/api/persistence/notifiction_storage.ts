import AsyncStorage from '@react-native-async-storage/async-storage'

const KEY = 'WALLET.NOTIFICATION'

export enum NotificationType {
  TRANSACTION = 'TRANSACTION',
}

export interface AppNotificationTypesI {
  type: NotificationType
  value: boolean
  displayName: string
}

function getAppNotificationTypes (): Array<{
  type: NotificationType
  displayName: string
  dafaultValue: boolean
}> {
  return [
    {
      type: NotificationType.TRANSACTION,
      displayName: 'Transactions',
      dafaultValue: true
    }
  ]
}

function notificationParser (preferances: AppNotificationTypesI[]): AppNotificationTypesI[] {
  const notificationsTypes = getAppNotificationTypes()
  return notificationsTypes.map(({ type, displayName, dafaultValue }) => {
    const savedType = preferances.find((each) => {
      return each.type === type
    })
    const value = (typeof savedType?.value === 'boolean') ? savedType?.value : dafaultValue
    return { type, displayName, value }
  }, {})
}

async function set (preferance: AppNotificationTypesI[]): Promise<void> {
  await AsyncStorage.setItem(KEY, JSON.stringify(preferance))
}

async function get (): Promise<AppNotificationTypesI[]> {
  const notificationTypes = await AsyncStorage.getItem(KEY) ?? '[]'
  return notificationParser(JSON.parse(notificationTypes))
}

export const NotificationPersistence = {
  set,
  get
}
