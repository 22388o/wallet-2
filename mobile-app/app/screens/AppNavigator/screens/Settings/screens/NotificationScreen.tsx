import { ThemedScrollView, ThemedText, ThemedView } from '@components/themed'
import { ThemedSectionTitle } from '@components/themed/ThemedSectionTitle'
import { Switch } from '@components/index'
import { tailwind } from '@tailwind'
import { translate } from '@translations'
import * as React from 'react'
import { useEffect, useState } from 'react'
import { View } from 'react-native'
import { Logging, NotificationPersistence } from '@api'
import { AppNotificationTypesI, NotificationType } from '@api/persistence/notifiction_storage'
import { WalletNotifications } from '@api/wallet/notifications'

export function NotificationScreen (): JSX.Element {
  const [notificationsTypes, setNotificationsTypes] = useState<AppNotificationTypesI[]>([])

  useEffect(() => {
    NotificationPersistence.get()
      .then((notificationsT) => {
        setNotificationsTypes(notificationsT)
      })
      .catch((err) => Logging.error(err))
  }, [])

  const onPreferanceChange = async (type: NotificationType, value: boolean): Promise<void> => {
    const updatedPreferance: AppNotificationTypesI[] = notificationsTypes.map((eachType) => {
      return type === eachType.type ? { ...eachType, value } : eachType
    })
    setNotificationsTypes(updatedPreferance)
    await NotificationPersistence.set(updatedPreferance)
    WalletNotifications.setAllowedNotificationTypes(updatedPreferance)
  }

  return (
    <ThemedScrollView light={tailwind('bg-white')}>
      <View testID='notification_screen'>
        <ThemedSectionTitle
          testID='notification_screen_title'
          text={translate('screens/NotificationScreen', 'APP NOTIFICATIONS')}
        />

        {notificationsTypes.map((eachType: AppNotificationTypesI) => (
          <RowNotificationItem key={eachType.type} item={eachType} onChange={onPreferanceChange} />
        ))}

        <View style={tailwind('p-4')}>
          <ThemedText
            dark={tailwind('text-gray-600')}
            light={tailwind('text-gray-400')}
            style={tailwind('text-xs font-normal')}
          >
            {translate('screens/NotificationScreen', 'Show a notification for your in-app transactions.')}
          </ThemedText>
        </View>
      </View>
    </ThemedScrollView>
  )
}

export const RowNotificationItem = (
  { item, onChange }: {item: AppNotificationTypesI, onChange: (type: NotificationType, value: boolean) => void}): JSX.Element => {
  return (
    <ThemedView
      dark={tailwind('bg-gray-800 border-b border-gray-700')}
      light={tailwind('bg-white border-b border-gray-200')}
      style={tailwind('flex flex-row p-4 pr-2 items-center justify-between')}
      testID='notification_type_row'
    >
      <ThemedText
        dark={tailwind('text-white text-opacity-90')}
        light={tailwind('text-black')}
        style={tailwind('font-medium')}
      >
        {translate('screens/NotificationScreen', item.displayName)}
      </ThemedText>

      <View style={tailwind('flex-row items-center')}>
        <Switch
          onValueChange={async (v) => {
            onChange(item.type, v)
          }}
          testID={`notification_${item.type}_switch`}
          value={item.value}
        />

      </View>
    </ThemedView>
  )
}
