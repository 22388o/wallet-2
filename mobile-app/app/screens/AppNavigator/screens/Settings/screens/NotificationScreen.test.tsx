import { AppNotificationTypesI, NotificationType } from '@api/persistence/notifiction_storage'
import { render } from '@testing-library/react-native'
import * as React from 'react'
import { NotificationScreen, RowNotificationItem } from './NotificationScreen'

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn()
}))

jest.mock('@api', () => ({
  NotificationPersistence: {
    get: async () => {
      return [{
        type: 'TRANSACTION',
        displayName: 'Transactions',
        value: true
      }]
    }
  }
}))

jest.mock('../../../../../contexts/ThemeProvider')

describe('Notification screen', () => {
  it('should render', async () => {
    const rendered = render(<NotificationScreen />)

    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('should render row notification item', async () => {
    const item: AppNotificationTypesI = {
      type: NotificationType.TRANSACTION,
      displayName: 'Transactions',
      value: true
    }
    const onChange = jest.fn
    const rendered = render(<RowNotificationItem item={item} onChange={onChange} />)

    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
