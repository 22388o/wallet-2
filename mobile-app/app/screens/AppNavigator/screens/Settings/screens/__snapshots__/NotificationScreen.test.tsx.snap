// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification screen should render 1`] = `
<RCTScrollView
  style={
    Array [
      undefined,
      Object {
        "backgroundColor": "rgba(255, 255, 255, 1)",
      },
    ]
  }
>
  <View>
    <View
      testID="notification_screen"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "RegularFont",
              "fontSize": 16,
              "fontWeight": "400",
              "lineHeight": 24,
            },
            Array [
              Object {
                "color": "rgba(115, 115, 115, 1)",
                "fontFamily": "MediumFont",
                "fontSize": 12,
                "fontWeight": "500",
                "lineHeight": 16,
                "paddingBottom": 16,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 24,
              },
              Object {
                "backgroundColor": "rgba(245, 245, 245, 1)",
                "color": "rgba(115, 115, 115, 1)",
              },
            ],
          ]
        }
        testID="notification_screen_title"
        text="APP NOTIFICATIONS"
      >
        APP NOTIFICATIONS
      </Text>
      <View
        style={
          Object {
            "paddingBottom": 16,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 16,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "RegularFont",
                "fontSize": 16,
                "fontWeight": "400",
                "lineHeight": 24,
              },
              Array [
                Object {
                  "fontFamily": "RegularFont",
                  "fontSize": 12,
                  "fontWeight": "400",
                  "lineHeight": 16,
                },
                Object {
                  "color": "rgba(163, 163, 163, 1)",
                },
              ],
            ]
          }
        >
          Show a notification for your in-app transactions.
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Notification screen should render row notification item 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 8,
        "paddingTop": 16,
      },
      Object {
        "backgroundColor": "rgba(255, 255, 255, 1)",
        "borderBottomColor": "rgba(229, 229, 229, 1)",
        "borderBottomWidth": 1,
        "borderLeftColor": "rgba(229, 229, 229, 1)",
        "borderRightColor": "rgba(229, 229, 229, 1)",
        "borderTopColor": "rgba(229, 229, 229, 1)",
      },
    ]
  }
  testID="notification_type_row"
>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "RegularFont",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        Array [
          Object {
            "fontFamily": "MediumFont",
            "fontWeight": "500",
          },
          Object {
            "color": "rgba(0, 0, 0, 1)",
          },
        ],
      ]
    }
  >
    Transactions
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <RCTSwitch
      accessibilityRole="switch"
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onTintColor="#34C759"
      style={
        Array [
          Object {
            "height": 31,
            "width": 51,
          },
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 16,
          },
        ]
      }
      testID="notification_TRANSACTION_switch"
      thumbTintColor="#fff"
      tintColor="rgba(120, 120, 128, 0.20)"
      value={true}
    />
  </View>
</View>
`;
